File: mip_reference/dotnet-aspnet/Mip/Models/SearchRequest.cs
using System.Text.Json.Serialization;

namespace Mip.Models;

/// <summary>
/// Tracks member search requests (inbound and outbound).
/// </summary>
public class SearchRequest
{
    public static readonly string[] Statuses = { "PENDING", "APPROVED", "DECLINED" };

    public string SharedIdentifier { get; set; } = Guid.NewGuid().ToString();
    public string Direction { get; set; } = ""; // "inbound" or "outbound"
    public string TargetMipIdentifier { get; set; } = "";
    public string TargetOrg { get; set; } = "";
    public Dictionary<string, string?> SearchParams { get; set; } = new();
    public string? Notes { get; set; }
    public List<object> Documents { get; set; } = new();
    public string Status { get; set; } = "PENDING";
    public List<Dictionary<string, object?>> Matches { get; set; } = new();
    public string? DeclineReason { get; set; }
    public string CreatedAt { get; set; } = DateTime.UtcNow.ToString("o");

    [JsonIgnore]
    public bool IsPending => Status == "PENDING";
    [JsonIgnore]
    public bool IsApproved => Status == "APPROVED";
    [JsonIgnore]
    public bool IsDeclined => Status == "DECLINED";
    [JsonIgnore]
    public bool IsInbound => Direction == "inbound";
    [JsonIgnore]
    public bool IsOutbound => Direction == "outbound";

    public void Approve(List<Dictionary<string, object?>> matches)
    {
        Status = "APPROVED";
        Matches = matches;
    }

    public void Decline(string? reason = null)
    {
        Status = "DECLINED";
        DeclineReason = reason;
    }

    public string SearchDescription
    {
        get
        {
            if (SearchParams.TryGetValue("member_number", out var memberNum) && !string.IsNullOrEmpty(memberNum))
                return $"Member #{memberNum}";

            if (SearchParams.TryGetValue("first_name", out var firstName) &&
                SearchParams.TryGetValue("last_name", out var lastName) &&
                !string.IsNullOrEmpty(firstName) && !string.IsNullOrEmpty(lastName))
            {
                var name = $"{firstName} {lastName}";
                if (SearchParams.TryGetValue("birthdate", out var birthdate) && !string.IsNullOrEmpty(birthdate))
                    name += $" ({birthdate})";
                return name;
            }

            return "Unknown search";
        }
    }

    public Dictionary<string, object?> ToRequestPayload()
    {
        var payload = new Dictionary<string, object?>
        {
            ["shared_identifier"] = SharedIdentifier
        };

        if (SearchParams.TryGetValue("member_number", out var memberNum) && !string.IsNullOrEmpty(memberNum))
            payload["member_number"] = memberNum;
        if (SearchParams.TryGetValue("first_name", out var firstName) && !string.IsNullOrEmpty(firstName))
            payload["first_name"] = firstName;
        if (SearchParams.TryGetValue("last_name", out var lastName) && !string.IsNullOrEmpty(lastName))
            payload["last_name"] = lastName;
        if (SearchParams.TryGetValue("birthdate", out var birthdate) && !string.IsNullOrEmpty(birthdate))
            payload["birthdate"] = birthdate;
        if (!string.IsNullOrEmpty(Notes))
            payload["notes"] = Notes;
        if (Documents.Count > 0)
            payload["documents"] = Documents;

        return payload;
    }

    public Dictionary<string, object?> ToReplyPayload()
    {
        return new Dictionary<string, object?>
        {
            ["shared_identifier"] = SharedIdentifier,
            ["status"] = Status,
            ["matches"] = Matches
        };
    }

    /// <summary>
    /// Create from received request payload.
    /// </summary>
    public static SearchRequest FromRequest(Dictionary<string, object?> payload, string senderMipId, string senderOrg)
    {
        var searchParams = new Dictionary<string, string?>();

        if (payload.TryGetValue("member_number", out var memberNum) && memberNum != null)
            searchParams["member_number"] = memberNum.ToString();
        if (payload.TryGetValue("first_name", out var firstName) && firstName != null)
            searchParams["first_name"] = firstName.ToString();
        if (payload.TryGetValue("last_name", out var lastName) && lastName != null)
            searchParams["last_name"] = lastName.ToString();
        if (payload.TryGetValue("birthdate", out var birthdate) && birthdate != null)
            searchParams["birthdate"] = birthdate.ToString();

        return new SearchRequest
        {
            SharedIdentifier = payload.GetValueOrDefault("shared_identifier")?.ToString() ?? Guid.NewGuid().ToString(),
            Direction = "inbound",
            TargetMipIdentifier = senderMipId,
            TargetOrg = senderOrg,
            SearchParams = searchParams,
            Notes = payload.GetValueOrDefault("notes")?.ToString(),
            Documents = (payload.GetValueOrDefault("documents") as List<object>) ?? new List<object>()
        };
    }
}

File: mip_reference/node-express/lib/mip/models/search-request.js
'use strict';

const crypto = require('crypto');

class SearchRequest {
  constructor(attrs = {}) {
    this.sharedIdentifier = attrs.sharedIdentifier || crypto.randomUUID();
    this.direction = attrs.direction; // 'inbound' or 'outbound'
    this.targetMipIdentifier = attrs.targetMipIdentifier;
    this.targetOrg = attrs.targetOrg;
    this.searchParams = attrs.searchParams || {};
    this.notes = attrs.notes || null;
    this.documents = attrs.documents || [];
    this.status = attrs.status || 'PENDING';
    this.matches = attrs.matches || [];
    this.declineReason = attrs.declineReason || null;
    this.createdAt = attrs.createdAt || new Date().toISOString();
  }

  isPending() { return this.status === 'PENDING'; }
  isApproved() { return this.status === 'APPROVED'; }
  isDeclined() { return this.status === 'DECLINED'; }
  isInbound() { return this.direction === 'inbound'; }
  isOutbound() { return this.direction === 'outbound'; }

  approve(matches) {
    this.status = 'APPROVED';
    this.matches = matches;
  }

  decline(reason) {
    this.status = 'DECLINED';
    this.declineReason = reason || null;
  }

  searchDescription() {
    if (this.searchParams.member_number) {
      return `Member #${this.searchParams.member_number}`;
    } else if (this.searchParams.first_name && this.searchParams.last_name) {
      let name = `${this.searchParams.first_name} ${this.searchParams.last_name}`;
      if (this.searchParams.birthdate) {
        name += ` (${this.searchParams.birthdate})`;
      }
      return name;
    }
    return 'Unknown search';
  }

  toRequestPayload() {
    const payload = {
      shared_identifier: this.sharedIdentifier
    };
    if (this.searchParams.member_number) payload.member_number = this.searchParams.member_number;
    if (this.searchParams.first_name) payload.first_name = this.searchParams.first_name;
    if (this.searchParams.last_name) payload.last_name = this.searchParams.last_name;
    if (this.searchParams.birthdate) payload.birthdate = this.searchParams.birthdate;
    if (this.notes) payload.notes = this.notes;
    if (this.documents.length > 0) payload.documents = this.documents;
    return payload;
  }

  toReplyPayload() {
    return {
      shared_identifier: this.sharedIdentifier,
      status: this.status,
      matches: this.matches
    };
  }

  // Create from received request payload
  static fromRequest(payload, senderMipId, senderOrg) {
    const searchParams = {};
    if (payload.member_number) searchParams.member_number = payload.member_number;
    if (payload.first_name) searchParams.first_name = payload.first_name;
    if (payload.last_name) searchParams.last_name = payload.last_name;
    if (payload.birthdate) searchParams.birthdate = payload.birthdate;

    return new SearchRequest({
      sharedIdentifier: payload.shared_identifier,
      direction: 'inbound',
      targetMipIdentifier: senderMipId,
      targetOrg: senderOrg,
      searchParams,
      notes: payload.notes,
      documents: payload.documents || []
    });
  }
}

module.exports = SearchRequest;
